File: game.py
import pygame
import random
import sys
import numpy as np

# Initialize Pygame and its mixer
pygame.init()
pygame.mixer.init()

# Constants
SCREEN_WIDTH, SCREEN_HEIGHT = 800, 600
BALL_SIZE = 30
WIN_SCORE = 100
GREEN = (0, 128, 0)
WHITE = (255, 255, 255)
BLACK = (0, 0, 0)
LIGHT_BLUE = (173, 216, 255)  # Light blue for sky
DARK_BLUE = (0, 0, 139)  # Darker blue for sky transition
GRASS_GREEN = (34, 139, 34)  # Dark green for grass

# Setup display
screen = pygame.display.set_mode((SCREEN_WIDTH, SCREEN_HEIGHT))
pygame.display.set_caption('Catch the Soccer Balls')

# Fonts and messages
font = pygame.font.Font(None, 36)
catch_messages = ["Look at you, finally making that button work!", "Awesome!", "You got it!", "Nice moves!", "Spot on!"]
miss_messages = ["Whoops!", "Missed!", "Are you playing or just sightseeing?", "Do you need glasses? Because that wasn’t even close!", "I’d call that a pass, but I don’t think the ball agrees."]

# Game variables
balls = []
catcher_x = SCREEN_WIDTH // 2
catcher_width = 100
player_speed = 20
score = 0
miss_count = 0
game_over = False
fall_speed = 5
message = ""
message_display_time = 0

# Clock for framerate control
clock = pygame.time.Clock()
FPS = 60

def draw_text(text, font, color, surface, x, y):
    textobj = font.render(text, True, color)
    textrect = textobj.get_rect()
    textrect.center = (x, y)
    surface.blit(textobj, textrect)

def draw_soccer_net(surface, x, y, width, height):
    """Draw a simple soccer net using lines."""
    pygame.draw.rect(surface, GREEN, (x, y, width, height), 2)
    for i in range(10):
        pygame.draw.line(surface, GREEN, (x + i * (width // 10), y), (x + i * (width // 10), y + height), 2)

def draw_gradient_background(surface, top_color, bottom_color):
    """Draw a vertical gradient from top_color to bottom_color."""
    for y in range(SCREEN_HEIGHT):
        color = tuple(map(lambda start, end: int(start + (end - start) * y / SCREEN_HEIGHT), top_color, bottom_color))
        pygame.draw.line(surface, color, (0, y), (SCREEN_WIDTH, y))

def draw_grass(surface, grass_color, height):
    """Draw a simple grass field at the bottom of the screen."""
    pygame.draw.rect(surface, grass_color, (0, SCREEN_HEIGHT - height, SCREEN_WIDTH, height))

def show_start_screen():
    """Show start screen until any key is pressed."""
    start_screen = True
    while start_screen:
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                pygame.quit()
                sys.exit()
            if event.type == pygame.KEYDOWN:
                start_screen = False
        draw_gradient_background(screen, LIGHT_BLUE, DARK_BLUE)
        draw_grass(screen, GRASS_GREEN, 100)
        draw_text("Press any key to start.", font, WHITE, screen, SCREEN_WIDTH // 2, SCREEN_HEIGHT // 2)
        pygame.display.flip()

show_start_screen()  # Display start screen at the beginning

# Main game loop
running = True
while running:
    draw_gradient_background(screen, LIGHT_BLUE, DARK_BLUE)
    draw_grass(screen, GRASS_GREEN, 100)
    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            pygame.quit()
            sys.exit()
        if event.type == pygame.KEYDOWN:
            if game_over and event.key == pygame.K_SPACE:
                balls = []
                score = 0
                miss_count = 0
                game_over = False
                fall_speed = 5
                catcher_width = 100

    if not game_over:
        if random.randint(1, 100) < 3:
            balls.append([random.randint(0, SCREEN_WIDTH - BALL_SIZE), 0])
        keys = pygame.key.get_pressed()
        if keys[pygame.K_LEFT] and catcher_x > 0:
            catcher_x -= player_speed
        if keys[pygame.K_RIGHT] and catcher_x < SCREEN_WIDTH - catcher_width:
            catcher_x += player_speed
        for ball in balls:
            ball[1] += fall_speed
            if ball[1] > SCREEN_HEIGHT - 50:
                if catcher_x < ball[0] + BALL_SIZE and catcher_x + catcher_width > ball[0]:
                    score += 10
                    message = random.choice(catch_messages)
                    message_display_time = 90
                else:
                    miss_count += 1
                    message = random.choice(miss_messages)
                    message_display_time = 90
                if miss_count >= 3:
                    game_over = True
                    message = "Game Over! Press Space to Restart"
                balls.remove(ball)

    if message_display_time > 0:
        draw_text(message, font, WHITE, screen, SCREEN_WIDTH // 2, SCREEN_HEIGHT // 2)
        message_display_time -= 1
    draw_soccer_net(screen, catcher_x, SCREEN_HEIGHT - 50, catcher_width, 50)
    for ball in balls:
        pygame.draw.circle(screen, BLACK, (ball[0] + BALL_SIZE // 2, ball[1] + BALL_SIZE // 2), BALL_SIZE // 2)
    draw_text(f'Score: {score}', font, WHITE, screen, SCREEN_WIDTH // 2, 20)
    pygame.display.flip()
    clock.tick(FPS)

pygame.quit()
sys.exit()

File: game.js
const canvas = document.getElementById('gameCanvas');
const ctx = canvas.getContext('2d');
let balls = [];
let catcherX = canvas.width / 2;
const catcherWidth = 100;
let score = 0;
let misses = 0;  // Keep track of how many balls the player missed
let maxMisses = 3;
let gameOver = false;
let fallSpeed = 5;  // Initial fall speed
let gameInterval;
let demoMode = false;  // Flag to activate demo mode

const catchMessages = ["Look at you, finally making that button work!", "Awesome!", "You got it!", "Nice moves!", "Spot on!"];
const missMessages = ["Whoops!", "Missed!", "Are you playing or just sightseeing?", "Do you need glasses? Because that wasn’t even close!", "I’d call that a pass, but I don’t think the ball agrees."];
let messageTimeout;

// Start the game when the button is clicked
document.getElementById('startButton').addEventListener('click', function() {
    document.getElementById('startPage').style.display = 'none';
    document.getElementById('gamePage').style.display = 'block';

    // Check if we want demo mode for automatic gameplay
    const isDemo = confirm("Do you want to watch the demo mode?");
    demoMode = isDemo;  // Enable demo mode if the user agrees
    startGame();
});

// Restart the game when the restart button is clicked
document.getElementById('restartButton').addEventListener('click', function() {
    document.getElementById('restartButton').style.display = 'none';
    document.getElementById('gameOver').style.display = 'none';
    startGame();
});

function drawCatcher() {
    ctx.fillStyle = 'green';
    ctx.fillRect(catcherX, canvas.height - 50, catcherWidth, 10);
}

function createBall() {
    return {
        x: Math.random() * canvas.width,
        y: 0,
        size: 30
    };
}

function drawBalls() {
    for (let i = 0; i < balls.length; i++) {
        const ball = balls[i];
        ball.y += fallSpeed;
        ctx.beginPath();
        ctx.arc(ball.x, ball.y, ball.size / 2, 0, Math.PI * 2);
        ctx.fillStyle = 'black';
        ctx.fill();

        if (ball.y > canvas.height - 50) {
            if (ball.x > catcherX && ball.x < catcherX + catcherWidth) {
                score += 10;
                showMessage(catchMessages[Math.floor(Math.random() * catchMessages.length)], 'green');
                balls.splice(i, 1);
                document.getElementById('score').textContent = 'Score: ' + score;

                // Increase fall speed every 50 points
                if (score % 50 === 0) {
                    fallSpeed += 1;  // Increase difficulty by increasing the ball fall speed
                }
            } else {
                misses++;
                showMessage(missMessages[Math.floor(Math.random() * missMessages.length)], 'red');
                balls.splice(i, 1);
                document.getElementById('misses').textContent = `Misses: ${misses} / ${maxMisses}`;

                if (misses >= maxMisses) {
                    gameOver = true;
                    endGame();
                    clearInterval(gameInterval);
                }
            }
        }
    }
}

// Automated catcher movement in demo mode
function automatedMove() {
    if (balls.length > 0) {
        const closestBall = balls[0];  // Get the first ball in the list
        if (closestBall.x < catcherX && catcherX > 0) {
            catcherX -= 10;  // Move left towards the ball
        } else if (closestBall.x > catcherX + catcherWidth && catcherX < canvas.width - catcherWidth) {
            catcherX += 10;  // Move right towards the ball
        }
    }
}

// Update the game to call automatedMove if in demo mode
function updateGame() {
    ctx.clearRect(0, 0, canvas.width, canvas.height);
    drawCatcher();
    if (Math.random() < 0.03) {
        balls.push(createBall());
    }
    drawBalls();
    
    if (demoMode) {
        automatedMove();  // Automate catcher movement
    }
}

document.addEventListener('keydown', function(event) {
    if (!gameOver && !demoMode) {  // Only move manually if not in demo mode
        if (event.key === 'ArrowLeft' && catcherX > 0) {
            catcherX -= 20;
        }
        if (event.key === 'ArrowRight' && catcherX < canvas.width - catcherWidth) {
            catcherX += 20;
        }
    }
});

function startGame() {
    // Reset game variables
    balls = [];
    score = 0;
    misses = 0;  // Reset the miss count
    fallSpeed = 5;  // Reset the speed
    gameOver = false;
    document.getElementById('gameOver').style.display = 'none';
    document.getElementById('score').textContent = 'Score: 0';
    document.getElementById('misses').textContent = `Misses: 0 / ${maxMisses}`;
    document.getElementById('restartButton').style.display = 'none';
    
    // Clear any previous game loop
    clearInterval(gameInterval);
    
    // Start the game loop
    gameInterval = setInterval(updateGame, 30);
}

function endGame() {
    document.getElementById('gameOver').style.display = 'inline';
    document.getElementById('restartButton').style.display = 'inline';
}

function showMessage(message, color) {
    clearTimeout(messageTimeout);
    const messageDiv = document.getElementById('message');
    messageDiv.textContent = message;
    messageDiv.style.color = color;
    messageTimeout = setTimeout(() => {
        messageDiv.textContent = '';
    }, 3000);  // Display the message for 3 seconds
}
